services/addons/images/firmware_manager/firmware_manager.py

```diff
@@ -0,0 +1,194 @@
+    return jsonify({"error": "Missing 'rsu_ip' parameter"}), 400
+
+  # Check if an upgrade is already occurring for the device
+  logging.info(f"Checking if existing upgrade is running for '{request_args['rsu_ip']}'")
+  if request_args['rsu_ip'] in active_upgrades:
+    return jsonify({"error": f"Firmware upgrade failed to start for '{request_args['rsu_ip']}': an upgrade is already underway for the target device"}), 500
+
+  # Pull RSU data from the PostgreSQL database
+  logging.info(f"Querying RSU data for '{request_args['rsu_ip']}'")
+  rsu_to_upgrade = get_rsu_upgrade_data(request_args['rsu_ip'])
+  if len(rsu_to_upgrade) == 0:
+    return jsonify({"error": f"Firmware upgrade failed to start for '{request_args['rsu_ip']}': the target firmware is already installed or is an invalid upgrade from the current firmware"}), 500
+  rsu_to_upgrade = rsu_to_upgrade[0]
+
+  # Start upgrade process
+  logging.info(f"Initializing firmware upgrade for '{request_args['rsu_ip']}'")
+  try:
+    p = Popen(['python3', f'/home/{manufacturer_upgrade_scripts[rsu_to_upgrade["manufacturer"]]}', f"'{json.dumps(rsu_to_upgrade)}'"], stdout=DEVNULL)
+    rsu_to_upgrade['process'] = p
+  except Exception as err:
+    logging.error(f"Encountered error of type {type(err)} while starting automatic upgrade process for {request_args['rsu_ip']}: {err}")
+    return jsonify({"error": f"Firmware upgrade failed to start for '{request_args['rsu_ip']}': upgrade process failed to run"}), 500
+
+  # Remove redundant ipv4_address from rsu_to_upgrade since it is the key for active_upgrades
+  del rsu_to_upgrade['ipv4_address']
+  active_upgrades[request_args['rsu_ip']] = rsu_to_upgrade
+  return jsonify({"message": f"Firmware upgrade started successfully for '{request_args['rsu_ip']}'"}), 201
+
+
+# REST endpoint to mark a firmware upgrade as complete and remove it from the active upgrades.
+# Will modify the RSU's firmware_version according to the "status" argument in the PostgreSQL database.
+# Required request body key-values:
+# - rsu_ip: Target device IPv4 address
+# - status: "success" or "fail" depending upon result of the firmware upgrade
+@app.route("/firmware_upgrade_completed", methods=["POST"])
+def firmware_upgrade_completed():
+  request_args = request.get_json()
+  if "rsu_ip" not in request_args:
+    return jsonify({"error": "Missing 'rsu_ip' parameter"}), 400
+  elif request_args['rsu_ip'] not in active_upgrades:
+    return jsonify({"error": "Specified device is not actively being upgraded or was already completed"}), 400
+
+  if "status" not in request_args:
+    return jsonify({"error": "Missing 'status' parameter"}), 400
+  elif request_args['status'] != "success" and request_args['status'] != "fail":
+    return jsonify({"error": "Wrong value for 'status' parameter - must be either 'success' or 'fail'"}), 400
+
+  # Update RSU firmware_version in PostgreSQL if the upgrade was successful
+  if request_args['status'] == "success":

```

Comment:
Do we need a case for the "fail" to send a notice that this failed somewhere? Maybe a separate PR to handle notifying support on a fail case so we can look into it and see if the whole RSU is down or otherwise troubleshoot?

Reply:
Yeah, some sort of email would definitely be useful. Right now it will continue to try to perform the upgrade every hour until it succeeds if it fails on the first attempt.